### ISSUE TITLE
Multiple "Begin Log" messages when running application multithreaded with verbose logging

### ISSUE THREAD

**Issue by Niilis** [jazzband/django-axes#312]
Using this app with the verbose logging is what I need for functionality. There is, however, an undesirable behavior with the three "Begin Log" messages and servers running in multiple threads.

Each time the server is started with AXES_VERBOSE, Axes sends the three messages (example below) to the logs. While this does not cause that much of a problem with servers running only in one thread, it is more of an issue with servers starting multiple threads of that app with Django-Axes installed.

`[01/Mar/2018 16:16:33] INFO [login:18] AXES: BEGIN LOG`
`[01/Mar/2018 16:16:33] INFO [login:19] AXES: Using django-axes 4.1.0`
`[01/Mar/2018 16:16:33] INFO [login:23] AXES: blocking by combination of username and IP.`

For each started app, Axes is filling the logs with the three "Begin Log messages" (once for each opened thread). With, let's say, 4 instances started, it would log 4 sets of messages in total.

To keep the verbose logging of login detail, my solution is to add a boolean to the if statement with a new parameter "disable begin log messages".

My question is then, has anyone else found this behavior? Or would there be a better solution for this?

**Comment by aleksihakli** [comment]
Hey, and thanks for reporting this @Niilis!

The cause of that logging is some misplaced logging code in [here](https://github.com/jazzband/django-axes/blob/9ecbf4cf63a361ed1f4b03e4683ca2356033f971/axes/decorators.py).

If you are inclined to improve the current functionality, you could try moving the logging into [`axes.apps.AppConfig.ready` method](https://github.com/jazzband/django-axes/blob/9ecbf4cf63a361ed1f4b03e4683ca2356033f971/axes/apps.py). [`ready` should only get called once per application instantiation in Django](https://docs.djangoproject.com/en/2.0/ref/applications/#django.apps.AppConfig.ready). 

Open a PR with the improvements and we'll merge it into mainline and make a new release :)

**Comment by Niilis** [comment]
Hi, and thank you for a rapid answer!

I have investigated this further, and unfortunately the fix you are suggesting is not working. I tried moving the message logging to the AppConfig.ready function, but the problem is still present.

When the app is loaded, i.e. with `./manage.py runserver`, you will have duplicate log messages as [this discussion](https://stackoverflow.com/q/33814615) also shows. A working solution would need more work, maybe toward [this answer](https://stackoverflow.com/a/47114949)?

**Comment by aleksihakli** [comment]
If you are running with runserver then it's possible that some logic runs multiple times when the server reloads. However, this should not be an issue in production setups with Gunicorn or uwsgi or the like.

Is this only occurring with your development setup with runserver or with your production setup as well?If it is occurring with production, could you supply your production server startup?

**Comment by Niilis** [comment]
This was occurring in both the development and the production environment. But, I have gotten the fix to work in production by specifying at what stage the app forking is happening.

With Gunicorn, it is possible to specify the app to be preloaded before forking the process into the desired amount of processes. By preloading, the logging happens only once. Without preloading, multiple logging is still present (with the fix or not).

With uwsgi, I have not gotten a definite answer, but I believe the forking is done after loading the app. Hence, this problem is not present.

**Comment by aleksihakli** [comment]
Nice to know! We'll have to see how to solve the multiple logging in `runserver` or if it is possible to mitigate the multiple logging in global scope at all. I'd imagine that it should be possible to adjust the logging but it would be nice to not to resort any hackyish approaches here.

I think that the forking is eager if it is done before the applications are loaded into memory, but the apps load lazily if they are loaded into memory only when a process is needed and instantiated?

**Comment by Niilis** [comment]
[This article](https://engineering.ticketea.com/uwsgi-preforking-lazy-apps/) explains the forking process quite well in my opinion, especially if you are unfamiliar with the process (myself included).

I agree with you on avoiding hackyish solutions. It might be hard to implement a universal solution with that kind of an approach. Unfortunately, I am not experienced enough to suggest or find an answer to this problem. Through the little research I made, many seemed to have difficulties with any kind of printing in the AppConfig.ready function. On the Django pages they talk about using flags for tasks that needs to be performed only once. I did not, however, find any references to example usage of such flags.

Thank you for the responses so far! Hope there is a solution to this problem.
